### PR TITLE
Add HTTP POST capability to load_data()

### PR DESCRIPTION
**Pull request by mtolk**
**IMPORTANT: Please do not create a Pull Request adding a new feature without discussing it first.**

The place to discuss new features is the forum: <https://zola.discourse.group/>
If you want to add a new feature, please open a thread there first in the feature requests section.

Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes
(Delete or ignore this section for documentation changes)

* [x] Are you doing the PR on the `next` branch?

If the change is a new feature or adding to/changing an existing one:

* [x] Have you created/updated the relevant documentation page(s)?

This PR adds POST capability to the load_data function.

Three optional parameters are added:
- `method` value must be either GET or POST. setting this to POST will make the request a POST request.
- `body` can be used to send a body in the POST request.
- `contenttype` sets the content-type header and should reflect the mime-type of the body.

These parameters are ignored when path is used instead of url.

### CODE REVIEW

**Comment by mtolk** [comment]
@Keats Hi When working with the functionality added previously in this PR i noticed i forgot something.

The new new parameter needed to be taken into account when hashing for the cache. 
I attempted to add this functionality. 

Tests are added.

I request you review these changes. Im not that exerienced in Rust yet. I was having some problems with the borrow checker,
so the resulting code may not be optimal.

Hope this endsup in the next release.

File: components/templates/src/global_fns/load_data.rs [hunk 1]
```diff
@@ -201,6 +207,19 @@ impl TeraFn for LoadData {
         };
         let path_arg = optional_arg!(String, args.get("path"), GET_DATA_ARGUMENT_ERROR_MESSAGE);
         let url_arg = optional_arg!(String, args.get("url"), GET_DATA_ARGUMENT_ERROR_MESSAGE);
+		let post_body_arg = optional_arg!(String, args.get("body"), "`load_data` body must be a string, if set.");
+		let post_content_type = optional_arg!(String, args.get("contenttype"), "`load_data` contenttype must be a string, if set.");
```
Comment (Keats) [inline components/templates/src/global_fns/load_data.rs]: let's do `content_type`

File: components/templates/src/global_fns/load_data.rs [hunk 2]
```diff
@@ -23,6 +24,11 @@ enum DataSource {
     Path(PathBuf),
 }
 
+enum Method {
```
Comment (Keats) [inline components/templates/src/global_fns/load_data.rs:28]: For enums like that, derive automatically the basic Debug/PartialEq/Clone/Copy it's almost free

File: components/templates/src/global_fns/load_data.rs [hunk 3]
```diff
@@ -201,6 +207,19 @@ impl TeraFn for LoadData {
...
+		let post_body_arg = optional_arg!(String, args.get("body"), "`load_data` body must be a string, if set.");
+		let post_content_type = optional_arg!(String, args.get("contenttype"), "`load_data` contenttype must be a string, if set.");
+		let method_arg = optional_arg!(String, args.get("method"), "`load_data` method must either be POST or GET.");
+		let method = match method_arg {
+			Some(method_str) => {
+				if method_str.to_lowercase() == "post"{
+				Method::Post
+				} else if method_str.to_lowercase() == "get" {
+				Method::Get
+				} else {return Err("`load_data` method must either be POST or GET.".into())}
```
Comment (Keats) [inline components/templates/src/global_fns/load_data.rs]: Not worth it here, but there is the https://doc.rust-lang.org/std/str/trait.FromStr.html for that case in general

File: docs/content/documentation/templates/overview.md [hunk 4]
```diff
@@ -339,6 +339,18 @@ as below.
...
 
+When no other parameters are specified the URL will always be retrieved using a HTTP GET request.
+Using the parameter `method`, since version 0.14.0, you can also choose to retrieve the URL using a POST request.
+
+When using `method="POST"` you can also use the parameters `body` and `contenttype`.
+The parameter body is the actual contents sent in the POST request.
+The parameter contenttype should be the mimetype of the body.
+
+```jinja2
+{% set postresponse = load_data(url="https://postman-echo.com/post", format="json", method="POST", body="{i_am:'json'}", contenttype="application/json") %}
```
Comment (Keats) [inline docs/content/documentation/templates/overview.md]: That's not valid JSON for the body. Can we get some real example?

**Review by Keats — CHANGES_REQUESTED** [review]
You should install rustfmt to solve the formatting issues

File: docs/content/documentation/templates/overview.md (hunk 4 above)
Comment (mtolk) [reply docs/content/documentation/templates/overview.md]: i commited a new example calling the kroki.io service to generate an SVG

File: components/templates/src/global_fns/load_data.rs (hunk 3 above)
Comment (mtolk) [reply components/templates/src/global_fns/load_data.rs]: implemented in the latest commit

File: components/templates/src/global_fns/load_data.rs (hunk 2 above)
Comment (mtolk) [reply components/templates/src/global_fns/load_data.rs:28]: implemented in the latest commit

File: components/templates/src/global_fns/load_data.rs (hunk 1 above)
Comment (mtolk) [reply components/templates/src/global_fns/load_data.rs]: done

**Comment by mtolk** [comment]
Hi @Keats. Thanks for all the feedback, it helps me learning Rust. Hopefully this also helps the Zola project.

File: components/templates/src/global_fns/load_data.rs [hunk 5]
```diff
@@ -23,6 +24,26 @@ enum DataSource {
...
+
+    fn from_str(s: &str) -> Result<Self> {
+        if s.to_lowercase() == "post" {
+            Ok(Method::Post)
+        } else if s.to_lowercase() == "get" {
+            Ok(Method::Get)
+        } else {
+            return Err("`load_data` method must either be POST or GET.".into());
+        }
+    }
```
Comment (Keats) [inline components/templates/src/global_fns/load_data.rs:42]:
```rust
match s.lowercase() {
  "post" => Ok(Method::Post),
  "get" => Ok(Method::Get),
  _ => Err("`load_data` method must either be POST or GET.".into())
}
```
would be more idiomatic

**Review by Keats — APPROVED** [review]
Thanks! I'll merge it soon

File: components/templates/src/global_fns/load_data.rs (hunk 5 above)
Comment (mtolk) [reply components/templates/src/global_fns/load_data.rs:42]: just commited the change to make the suggested change

**Comment by mtolk** [comment]
sorry misclicked, I did not mean to request a review. Sorry

**Review by Keats — APPROVED** [review]

**Comment by Keats** [comment]
Thanks!

**Pull request merged**